### ISSUE TITLE
Hyprland v0.40 breaks fcitx with keys getting stuck

### ISSUE THREAD

**Issue by acheong08** [hyprwm/Hyprland#5880]
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch main at commit 1237732b97fcf21a19aed16f2ef966118a3ca2aa  (input: Introduce basic hyprland HID classes).
Date: Fri May 3 22:40:27 2024
Tag: v0.39.1-139-g1237732b, commits: 4599

flags: (if any)


System Information:
System name: Linux
Node name: fish
Release: 6.8.8-300.fc40.x86_64
Version: #1 SMP PREEMPT_DYNAMIC Sat Apr 27 17:53:31 UTC 2024


GPU information: 
00:02.0 VGA compatible controller [0300]: Intel Corporation TigerLake-LP GT2 [Iris Xe Graphics] [8086:9a49] (rev 01) (prog-if 00 [VGA controller])


os-release: NAME="Fedora Linux"
VERSION="40 (Workstation Edition)"
```

</details>

### Bug or Regression?

Bug

### Description

When typing, the control key sometimes randomly activates. I have to press the control key again to disable it. Unlikely to be hardware as I rolled back to the previous v0.38.1 and it worked fine.

Edit: Rolled back to v0.39.1 and everything is working nicely as well.

### How to reproduce

Type stuff and wait

### Crash reports, logs, images, videos

I don't see anything relevant in the logs

**Comment by vaxerski** [comment]
`git bisect` then please

**Comment by acheong08** [comment]
Takes forever to compile... Seems to have fixed itself on the latest commit. Closing.

**Comment by acheong08** [comment]
Oh I was reproducing it wrong. Now able to reproduce by just holding control for really long, then pressing a button for a while along side. After releasing, it gets stuck when I press other keys. Just finished git bisecting another issue. Will do that now

**Comment by acheong08** [comment]
The reason for the weird `git bisect` is because I forgot to update submodules each time, only when build fails. The issue is likely coming from a submodule.

**Comment by acheong08** [comment]
This issue is making v0.40.0 near unusable. Staying on 4ed6b69b68a544a81cec583a1ad8aff218f7db5d with cherry picked cddeec47a1fc0e70d8598fd10c29cd0e9489999f. I'm surprised nobody else is affected by this

**Comment by vaxerski** [comment]
are you using fcitx / ibus?

**Comment by acheong08** [comment]
Yes. I use English pretty much 99% of the time though. Will test with fcitx uninstalled.

**Comment by nergdron** [comment]
I'm running into similar behaviour since upgrading to v0.40 as well. however what happens to me is that if I hit a control sequence (such as ctl-pgdown, which I do a lot to switch through tabs in apps), the first one or two presses work, and then it removes the ctl modifier and starts just doing the raw keypress. this breaks in the same way in every app I can test, such as hitting ctl-c in a terminal, the first 1-2 work, and then I just get a bunch of lowercase Cs fed to the terminal.

I'm also using fcitx, so that seems like common issues. I'll test shortly without fcitx starting in my session, and report back.

**Comment by acheong08** [comment]
Yup, killing fcitx fixed it for me as well. https://github.com/hyprwm/Hyprland/commit/8bcccf9f0f0f69986e315d4288dfdbf119dd19db is 100% the commit that introduced this bug

**Comment by 15cm** [comment]
Didn't look deep into the issue but I hit the same problem "ctrl modifier doesn't work unless I release the ctrl key and press it again". It only happens in the combination of:
- In Firefox web browser
- Fcitx5 is running
- After I upgraded Hyprland from commit https://github.com/hyprwm/Hyprland/commit/79a139c9495568f69dd995bce1ca579247a98a17 to commit https://github.com/hyprwm/Hyprland/commit/f15513309b24790099d42974274eb23f66f7c985

All other applications like Chrome, Emacs, Alacritty... all work fine.

Killing Fcitx5 fixes the issue for me as well.

**Comment by acheong08** [comment]
> In Firefox web browser

I reproduced it in `foot` as well. It's definitely not a firefox bug

**Comment by 15cm** [comment]
Yeah I meant for narrowing down the scenario of the issue.

I tried both Hyprland Nix flake built on top of https://github.com/hyprwm/Hyprland/commit/8bcccf9f0f0f69986e315d4288dfdbf119dd19db and its parent https://github.com/hyprwm/Hyprland/commit/4ed6b69b68a544a81cec583a1ad8aff218f7db5d. I can consistently reproduce in the former and cannot in the latter.

**Comment by 15cm** [comment]
The issue goes away after I:
- Bumped hyprland to https://github.com/hyprwm/Hyprland/commit/bf75723f2742973d4820f3f5378dff8c99333660
- Changed GTK_IM_MODULE from `wayland` to unset following the latest recommendation of https://fcitx-im.org/wiki/Using_Fcitx_5_on_Wayland

**Comment by acheong08** [comment]
@15cm Are you sure? I've done that and the issue still pops up from time to time

**Comment by 15cm** [comment]
Yes. I have been not encountering the issue in Firefox since then. IIRC I also upgraded fcitx5 when I posted.

Here is the whole configuration (package versions, config files …) of my machines in case it helps: https://github.com/15cm/nixfiles

**Comment by 15cm** [comment]
After a few more days of using, I find that the issue can still be reproduced occasionally, but just not as frequent as before. Agree with @acheong08 that it doesn't completely go away.

**Comment by Zhaith-Izaliel** [comment]
> Yes. I have been not encountering the issue in Firefox since then. IIRC I also upgraded fcitx5 when I posted.
> 
> Here is the whole configuration (package versions, config files …) of my machines in case it helps: https://github.com/15cm/nixfiles

Thank you @15cm, your configuration help me workaround this issue in Hyprland 0.41.2 !

**Comment by acheong08** [comment]
It's been fixed
